File: src/algebra2_rational_equations_hard.py
"""
algebra2 - rational_equations (hard)
Generated: 2026-02-22T04:25:21.418605
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: Rational equation with multiple fractions requiring LCD
        # Harder version: 3 fractions with different denominators
        a1 = nonzero(-5, 5)
        b1 = nonzero(-8, 8)
        a2 = nonzero(-5, 5)
        b2 = nonzero(-8, 8)
        a3 = nonzero(-5, 5)
        b3 = nonzero(-8, 8)
        
        # Create denominators that share factors for interesting LCD
        d1 = nonzero(2, 4)
        d2 = nonzero(2, 4)
        while d2 == d1:
            d2 = nonzero(2, 4)
        d3 = d1 * d2
        
        # Equation: a1/(d1(x-b1)) + a2/(d2(x-b2)) = a3/(d3(x-b3))
        eq = Eq(a1/(d1*(x - b1)) + a2/(d2*(x - b2)), a3/(d3*(x - b3)))
        
        sols = solve(eq, x)
        # Filter out solutions that make denominators zero
        valid_sols = [s for s in sols if s != b1 and s != b2 and s != b3 and s.is_real]
        
        if len(valid_sols) == 0:
            # Fallback to simpler problem
            a1, b1, a2, b2 = 2, 3, 1, 5
            d1, d2 = 1, 1
            eq = Eq(a1/(x - b1) + a2/(x - b2), 1)
            sols = solve(eq, x)
            valid_sols = [s for s in sols if s != b1 and s != b2 and s.is_real]
        
        ans = valid_sols[0] if len(valid_sols) > 0 else 0
        
        return problem(
            question=f"Solve for $x$: ${latex(eq.lhs)} = {latex(eq.rhs)}$",
            answer=ans,
            difficulty=(1650, 1750),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Given: ${latex(eq)}$",
                f"Multiply both sides by the LCD of all denominators",
                f"Simplify and solve the resulting equation",
                f"Check for extraneous solutions (values that make denominators zero)",
                f"$x = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 2:
        # Type 2: Rational equation leading to quadratic
        # (x+a)/(x-b) = c/(x-d) leads to quadratic after cross-multiplying
        a_val = nonzero(-6, 6)
        b_val = nonzero(-6, 6)
        c_val = nonzero(-6, 6)
        d_val = nonzero(-6, 6)
        
        while b_val == d_val:
            d_val = nonzero(-6, 6)
        
        eq = Eq((x + a_val)/(x - b_val), c_val/(x - d_val))
        
        sols = solve(eq, x)
        valid_sols = [s for s in sols if s != b_val and s != d_val and s.is_real]
        
        if len(valid_sols) == 0:
            # Fallback
            a_val, b_val, c_val, d_val = 3, 2, 4, -1
            eq = Eq((x + a_val)/(x - b_val), c_val/(x - d_val))
            sols = solve(eq, x)
            valid_sols = [s for s in sols if s != b_val and s != d_val and s.is_real]
        
        if len(valid_sols) == 1:
            ans = valid_sols[0]
        else:
            ans = FiniteSet(*valid_sols)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq.lhs)} = {latex(eq.rhs)}$",
            answer=ans,
            difficulty=(1700, 1800),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Given: ${latex(eq)}$",
                f"Cross-multiply: ${latex((x + a_val)*(x - d_val))} = {latex(c_val*(x - b_val))}$",
                f"Expand and simplify to get a quadratic equation",
                f"Solve the quadratic and check for extraneous solutions",
                f"Solution: ${latex(ans)}$"
            ),
        )
    
    elif problem_type == 3:
        # Type 3: Work/rate problem leading to rational equation
        # Person A takes a hours, Person B takes b hours, together they take t hours
        # 1/a + 1/b = 1/t, solve for one variable
        a_time = randint(4, 12)
        together_time = randint(2, a_time - 1)
        
        # Find b such that 1/a + 1/b = 1/t
        # 1/b = 1/t - 1/a = (a - t)/(a*t)
        # b = a*t/(a - t)
        
        eq = Eq(1/a_time + 1/x, 1/together_time)
        sols = solve(eq, x)
        valid_sols = [s for s in sols if s.is_real and s > 0]
        
        if len(valid_sols) == 0 or not valid_sols[0].is_rational:
            # Pick values that work nicely
            a_time = 6
            together_time = 2
            eq = Eq(1/a_time + 1/x, 1/together_time)
            sols = solve(eq, x)
            valid_sols = [s for s in sols if s.is_real and s > 0]
        
        ans = valid_sols[0]
        
        return problem(
            question=f"Machine A can complete a job in ${a_time}$ hours. When Machine A and Machine B work together, they can complete the job in ${together_time}$ hours. How many hours would it take Machine B to complete the job alone?",
            answer=ans,
            difficulty=(1600, 1700),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Machine A's rate: $\\frac{{1}}{{{a_time}}}$ of the job per hour",
                f"Combined rate: $\\frac{{1}}{{{together_time}}}$ of the job per hour",
                f"Let $x$ = hours for Machine B alone, so Machine B's rate is $\\frac{{1}}{{x}}$",
                f"Set up equation: $\\frac{{1}}{{{a_time}}} + \\frac{{1}}{{x}} = \\frac{{1}}{{{together_time}}}$",
                f"Solve: $\\frac{{1}}{{x}} = \\frac{{1}}{{{together_time}}} - \\frac{{1}}{{{a_time}}}$",
                f"$x = {latex(ans)}$ hours"
            ),
        )
    
    else:
        # Type 4: Complex rational equation with variable in multiple places
        # (x+a)/(x+b) + (x+c)/(x+d) = k
        a_val = nonzero(-5, 5)
        b_val = nonzero(-5, 5)
        c_val = nonzero(-5, 5)
        d_val = nonzero(-5, 5)
        k_val = randint(2, 4)
        
        while b_val == d_val or abs(b_val - d_val) < 2:
            d_val = nonzero(-5, 5)
        
        eq = Eq((x + a_val)/(x + b_val) + (x + c_val)/(x + d_val), k_val)
        
        sols = solve(eq, x)
        valid_sols = [s for s in sols if s != -b_val and s != -d_val and s.is_real]
        
        if len(valid_sols) == 0:
            # Fallback
            a_val, b_val, c_val, d_val, k_val = 1, 2, 3, -1, 3
            eq = Eq((x + a_val)/(x + b_val) + (x + c_val)/(x + d_val), k_val)
            sols = solve(eq, x)
            valid_sols = [s for s in sols if s != -b_val and s != -d_val and s.is_real]
        
        if len(valid_sols) == 1:
            ans = valid_sols[0]
        else:
            ans = FiniteSet(*valid_sols)
        
        return problem(
            question=f"Solve for $x$: ${latex(eq.lhs)} = {latex(eq.rhs)}$",
            answer=ans,
            difficulty=(1750, 1850),
            topic="algebra2/rational_equations",
            solution=steps(
                f"Given: ${latex(eq)}$",
                f"Find LCD: $(x + {b_val})(x + {d_val})$",
                f"Multiply both sides by the LCD",
                f"Expand and collect terms to form a quadratic equation",
                f"Solve and verify solutions don't make denominators zero",
                f"Solution: ${latex(ans)}$"
            ),
        )

emit(generate())

File: src/algebra2_rational_equations_hard.jl
# algebra2 - rational_equations (hard)
# Generated: 2026-03-08T20:16:46.130987

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("algebra2/rational_equations")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Type 1: Basic rational equation with two fractions equal to each other
        # ELO 1800-2000
        a1 = nonzero(-8, 8)
        b1 = randint(-15, 15)
        a2 = nonzero(-8, 8)
        b2 = randint(-15, 15)
        
        # Denominators with linear factors
        den1_root = nonzero(-10, 10)
        den2_root = nonzero(-10, 10)
        while den2_root == den1_root
            den2_root = nonzero(-10, 10)
        end
        
        den1 = x - den1_root
        den2 = x - den2_root
        
        # Cross multiply to find solution
        # (a1*x + b1) / den1 = (a2*x + b2) / den2
        # (a1*x + b1) * den2 = (a2*x + b2) * den1
        lhs = expand((a1*x + b1) * den2)
        rhs = expand((a2*x + b2) * den1)
        
        sols = solve(lhs ~ rhs, x)
        if length(sols) > 0
            ans = sols[1]
            # Check it's not an excluded value
            if Float64(ans) != Float64(den1_root) && Float64(ans) != Float64(den2_root)
                num1 = a1*x + b1
                num2 = a2*x + b2
                
                problem(
                    question="Solve the rational equation: \\frac{$(tex(num1))}{$(tex(den1))} = \\frac{$(tex(num2))}{$(tex(den2))}",
                    answer=ans,
                    difficulty=(1800, 2000),
                    solution=steps(
                        sol("Given equation", num1/den1 ~ num2/den2),
                        "Cross multiply: \$($(tex(num1)))($(tex(den2))) = ($(tex(num2)))($(tex(den1)))\$",
                        sol("Expand", lhs ~ rhs),
                        "Solve for x",
                        sol("Answer", ans),
                        "Verify \\(x \\neq $(den1_root), $(den2_root)\\) (excluded values)"
                    ),
                    time=180
                )
            else
                # Solution is excluded, create no solution case
                problem(
                    question="Solve the rational equation: \\frac{$(tex(num1))}{$(tex(den1))} = \\frac{$(tex(num2))}{$(tex(den2))}",
                    answer="no solution",
                    difficulty=(2000, 2200),
                    answer_type="text",
                    solution=steps(
                        sol("Given equation", num1/den1 ~ num2/den2),
                        "Cross multiply and solve",
                        "Get \\(x = $(ans)\\)",
                        "But \\(x = $(ans)\\) makes a denominator zero",
                        "Therefore: no solution"
                    ),
                    time=200
                )
            end
        else
            # Backup simple problem
            ans = randint(-10, 10)
            a = nonzero(-5, 5)
            b = randint(-10, 10)
            den = x - nonzero(-8, 8)
            num = expand(a*(x - ans) + b)
            problem(
                question="Solve: \\frac{$(tex(num))}{$(tex(den))} = $(b)",
                answer=ans,
                difficulty=(1800, 1900),
                solution=steps(
                    "Multiply both sides by denominator",
                    sol("Simplified", num ~ b*den),
                    "Solve for x",
                    sol("Answer", ans)
                ),
                time=150
            )
        end
        
    elseif problem_type == 2
        # Type 2: Sum of rational expressions equals a constant
        # ELO 1900-2100
        den1_root = nonzero(-10, 10)
        den2_root = nonzero(-10, 10)
        while den2_root == den1_root
            den2_root = nonzero(-10, 10)
        end
        
        den1 = x - den1_root
        den2 = x - den2_root
        
        # Pick answer first
        ans = randint(-12, 12)
        while ans == den1_root || ans == den2_root
            ans = randint(-12, 12)
        end
        
        # Create numerators that give a nice equation
        a1 = nonzero(-6, 6)
        a2 = nonzero(-6, 6)
        c = randint(-10, 10)
        
        # a1/(x-r1) + a2/(x-r2) = c
        # Common denominator: (x-r1)(x-r2)
        # a1(x-r2) + a2(x-r1) = c(x-r1)(x-r2)
        
        lhs_num = expand(a1*(x - den2_root) + a2*(x - den1_root))
        rhs_expanded = expand(c * (x - den1_root) * (x - den2_root))
        
        problem(
            question="Solve the equation: \\frac{$(a1)}{$(tex(den1))} + \\frac{$(a2)}{$(tex(den2))} = $(c)",
            answer=ans,
            difficulty=(1900, 2100),
            solution=steps(
                "Find common denominator: \$($(tex(den1)))($(tex(den2)))\$",
                "Multiply both sides by LCD",
                sol("Equation becomes", lhs_num ~ rhs_expanded),
                "Simplify and solve for x",
                sol("Answer", ans),
                "Check excluded values \\(x \\neq $(den1_root), $(den2_root)\\)"
            ),
            time=210
        )
        
    elseif problem_type == 3
        # Type 3: Rational equation with quadratic denominator
        # ELO 2000-2300
        r1 = nonzero(-8, 8)
        r2 = nonzero(-8, 8)
        while r2 == r1
            r2 = nonzero(-8, 8)
        end
        
        quad_den = expand((x - r1) * (x - r2))
        linear_den = x - r1
        
        # Pick clean answer
        ans = randint(-10, 10)
        while ans == r1 || ans == r2
            ans = randint(-10, 10)
        end
        
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        c = randint(-8, 8)
        
        # a/(x-r1) + b/[(x-r1)(x-r2)] = c
        num1 = a
        num2 = b
        
        problem(
            question="Solve: \\frac{$(a)}{$(tex(linear_den))} + \\frac{$(b)}{$(tex(quad_den))} = $(c)",
            answer=ans,
            difficulty=(2000, 2300),
            solution=steps(
                "Factor denominators if needed",
                "LCD is \$($(tex(quad_den)))\$",
                "Multiply through by LCD",
                "Get: \\($(a)($(tex(x - r2))) + $(b) = $(c)($(tex(quad_den)))\\)",
                "Expand and solve",
                sol("Answer", ans)
            ),
            time=240
        )
        
    elseif problem_type == 4
        # Type 4: Proportion with quadratics
        # ELO 2100-2400
        q1 = rand_factorable(x; roots=(-8, 8))
        q2 = rand_factorable(x; roots=(-8, 8))
        
        a = nonzero(-6, 6)
        b = randint(-10, 10)
        
        # (ax + b) / q1.expr = constant / q2.expr
        k = nonzero(-5, 5)
        
        # Cross multiply: (ax+b)*q2.expr = k*q1.expr
        lhs = expand((a*x + b) * q2.expr)
        rhs = expand(k * q1.expr)
        
        sols = solve(lhs ~ rhs, x)
        
        if length(sols) > 0
            ans = sols[1]
            excluded = Set([q1.r1, q1.r2, q2.r1, q2.r2])
            
            if !(Float64(ans) in Set(Float64.(collect(excluded))))
                problem(
                    question="Solve: \\frac{$(tex(a*x + b))}{$(tex(q1.expr))} = \\frac{$(k)}{$(tex(q2.expr))}",
                    answer=ans,
                    difficulty=(2100, 2400),
                    solution=steps(
                        "Cross multiply",
                        sol("Get", lhs ~ rhs),
                        "Expand both sides and collect like terms",
                        "Solve the resulting equation",
                        sol("Answer", ans),
                        "Verify solution doesn't make any denominator zero"
                    ),
                    time=270
                )
            else
                problem(
                    question="Solve: \\frac{$(tex(a*x + b))}{$(tex(q1.expr))} = \\frac{$(k)}{$(tex(q2.expr))}",
                    answer="no solution",
                    difficulty=(2200, 2500),
                    answer_type="text",
                    solution=steps(
                        "Cross multiply",
                        "Expand and solve",
                        "Algebraic solution: \\(x = $(ans)\\)",
                        "But \\(x = $(ans)\\) is an excluded value",
                        "Final answer: no solution"
                    ),
                    time=280
                )
            end
        else
            # Fallback
            ans = randint(-8, 8)
            num = expand(2*(x - ans))
            den = expand((x - nonzero(-6, 6)) * (x - nonzero(-6, 6)))
            problem(
                question="Solve: \\frac{$(tex(num))}{$(tex(den))} = 0",
                answer=ans,
                difficulty=(2000, 2200),
                solution=steps(
                    "A fraction equals zero when numerator is zero",
                    sol("Set numerator to zero", num ~ 0),
                    "Solve for x",
                    sol("Answer", ans)
                ),
                time=180
            )
        end
        
    else
        # Type 5: Work/rate problem leading to rational equation
        # ELO 2200-2500
        t1 = randint(3, 15)
        t2 = randint(3, 15)
        while t2 == t1
            t2 = randint(3, 15)
        end
        
        # Working together time
        t_together = (t1 * t2) // (t1 + t2)
        ans = float(t_together)
        
        problem(
            question="Person A can complete a job in $(t1) hours. Person B can complete the same job in $(t2) hours. How long will it take them to complete the job working together? (Give exact answer as a fraction or decimal)",
            answer=ans,
            difficulty=(2200, 2500),
            solution=steps(
                "Rate of A: \\(\\frac{1}{$(t1)}\\) jobs/hour",
                "Rate of B: \\(\\frac{1}{$(t2)}\\) jobs/hour",
                "Combined rate: \\(\\frac{1}{$(t1)} + \\frac{1}{$(t2)} = \\frac{1}{t}\\)",
                "Solve: \\(\\frac{$(t1 + t2)}{$(t1 * t2)} = \\frac{1}{t}\\)",
                sol("Answer", t_together)
            ),
            time=240
        )
    end
end